.github/container/jax_nsys/python/jax_nsys/jax_nsys/analysis.py

```diff
@@ -3,13 +3,111 @@
+    mean_end_times = end_times.groupby(
+        ["ProgramId", "ProgramExecution", "ThunkIndex"]
+    ).agg("mean")
+    # For each collective + device, calculate the delta of the end time from the mean
+    end_time_skews = end_times - mean_end_times
+    device_skews = end_time_skews.groupby("Device")
+    median_device_skews = device_skews.agg("median")
+    # Apply these corrections to the device-side timestamps
+    for k in ["communication", "module", "thunk"]:
+        df = frames[k]
+        df["ProjStartNs"] -= median_device_skews
+        frames[k] = df
+    return frames, {
+        "collective_end_time_skews_ns": end_time_skews,
+        "device_corrections": median_device_skews,
+        "collective_size": max_collective_size,
+    }
+
+
+def apply_warmup_heuristics(
+    frames: dict[str, pd.DataFrame],
+) -> tuple[dict[str, pd.DataFrame], dict[str, pd.DataFrame]]:
+    """
+    Given a dictionary of data frames, as returned by `load_profiler_data`, use
+    heuristics to split the profile data into initialisation and steady state running.
+    The current approach is to assume everything is steady state if compilation was not
+    profiled, and if compilation *was* profiled then label the 0th execution as
+    initialisation and the 2nd and later ones as steady state operation, discarding one
+    execution in between.
+
+    Returns a tuple of:
+      dictionary of data frames like the input, with only initialisation (and compile)
+      dictionary of data frames like the input, with only steady state (and no compile)
+    """
+    assert "compile" in frames
+    # Which program IDs did we witness compilation of?
+    compilation_ids_seen = sorted(frames["compile"]["ProgramId"].unique())
+    # Generally the first execution of a module will be slower, but we don't know for
+    # sure if the profile being analysed included the whole runtime or was more
+    # selective. As a heuristic, we can skip the first two executions of modules that
+    # we saw the compilation of. The motivation for skipping the second executions is
+    # that with a typical structure like:
+    #
+    # for n in range(n_iterations):
+    #  preamble(n)
+    #  step_function(n) # involves collectives/synchronisation
+    #  postamble(n)
+    #
+    # then one-time costs (e.g. JIT compilation) of postamble(0) will affect when
+    # step_function(1) is actually launched, whereas step_function(2) and later are
+    # expected to launch closer to in lockstep across processes.
+    init = {"compile": frames["compile"]}
+    steady = {}
+    for k in ["communication", "thunk", "module"]:
+        if k not in frames:
+            continue
+        df = frames[k]
+        compile_mask = df.index.get_level_values("ProgramId").isin(compilation_ids_seen)
+        prog_exec_values = df.index.get_level_values("ProgramExecution")
+        init_mask = compile_mask & (prog_exec_values == 0)
+        steady_mask = ~compile_mask | (compile_mask & (prog_exec_values > 1))
```

Comment:
Nit: This should be equivalent to `~compile_mask | (prog_exec_values > 1)`, no?

Reply:
Well spotted, thanks.